### PR TITLE
GangaGUI - Views & Web CLI

### PR DESCRIPTION
**Pull request by varunbankar**
# Ganga GUI - Views & Web CLI

### Summary
This PR contains the code for Ganga GUI view routes like (dashboard, create, jobs, config, etc.) and web CLI implementation. This PR also contains a major restructuring of the project over the previous code. Earlier there was just a single-threaded weak Flask webserver which ran on a GangaThread in the Ganga thread pool allowing it to access all the GPI functions (while keeping the CLI running) when Ganga was running. In this PR the project was restructured to make use of two web servers in order to server the GUI to the user. The way it works is, there are two separate Flask applications, one is called the `internal`, and the second one is called the `gui`. The `internal` Flask instance uses a single-threaded weak web server and runs on the GangaThread in the Ganga thread pool. This Flask instance has access to all the Ganga GPI functions in the background and its only work is to act as a RESTful interface between Ganga and the `gui` Flask instance. The `internal` server is only accessible from the local machine and not over the network and it is meant for internal use by the `gui` server in order to get information from the Ganga and render web pages according to it. The communication between the `gui` server and the `internal` server occurs in JSON format. The `gui` Flask instance is completely decoupled from Ganga and it runs as a separate process on a Gunicorn server (production-ready server). 

The series of events that happens when Ganga is called with the `--webgui` flag or `start_gui` function is invoked within the Ganga CLI is as follows:
* User instructs Ganga to start the GUI.
* A `GangaGUIAdmin` user is created for the user to be able to access the GUI using its credentials.
* The `internal` Flask instance is started first on a GangaThread and is attached to the Ganga thread pool. The `internal` server listens to internal HTTP requests on a port specified by the user or `5000` by default. (inaccessible over the network)
* Afterwards, a subprocess invokes the `gui` Flask instance on a Gunicorn server. This `gui` server listens to outside HTTP requests on a port specified by the user or `5500` by default. This is accessible over the network and this is the server that is exposed to the outside world.
* The credentials and the `gui` server information is displayed to the user.
* The user then makes an HTTP request to `gui` server on the server IP address and specified GUI port.
* The `gui` server listens to the request made by the user and then depending on the type of the information requested by the user, the `gui` server makes an internal HTTP to the `internal` Flask server in order to get the information from Ganga.
* The `internal` server listens to the HTTP request made by the `gui` server and accordingly gets the requested information from Ganga and returns it back to the `gui` server in a JSON format.
* The `gui` server then parses the JSON response of the `internal` server and uses this information to render the web page that is to be displayed to the user. (or forwards the JSON as is in case of API calls to the `gui` server).
* The `gui` server serves rendered web page to the user and it is then displayed in the user's browser.

---

### About `internal` Web Server

The `internal` is used by the `gui` server in order to get information from Ganga. The `internal` server is a single-threaded Flask application running on a GangaThread. It implements the following APIs for internal use. Check documentation for more information regarding these APIs.

**Job API**
* `/internal/jobs/<int:job_id>` - `GET`
* `/internal/jobs/<int:job_id>/<attribute>` - `GET`
* `/internal/jobs/<int:job_id>/full-print` - `GET`
* `/internal/jobs/create` - `POST`
* `/internal/jobs/<int:job_id>/copy` - `PUT`
* `/internal/jobs/<int:job_id>/<action>` - `PUT`
* `/internal/jobs/<int:job_id>` - `DELETE`
* `/internal/jobs/<int:job_id>/export` - `GET`

**Subjobs API**
* `/internal/jobs/<int:job_id>/subjobs` - `GET`
* `/internal/jobs/<int:job_id>/subjobs/length` - `GET`
* `/internal/jobs/<int:job_id>/subjobs/<int:subjob_id>` - `GET`
* `/internal/jobs/<int:job_id>/subjobs/<int:subjob_id>/<attribute>` - `GET`
* `/internal/jobs/<int:job_id>/subjobs/<int:subjob_id>/copy` - `PUT`
* `/internal/jobs/<int:job_id>/subjobs/<int:subjob_id>/full-print` - `GET`

**Jobs API**
* `/internal/jobs` - `GET`
* `/internal/jobs/length` - `GET`
* `/internal/jobs/statistics` - `GET`
* `/internal/jobs/incomplete-ids` - `GET`
* `/internal/jobs/recent` - `GET`

**Config API**
* `/internal/config` - `GET`
* `/internal/config/<section>` - `GET`

**Templates API**
* `/internal/templates` - `GET`
* `/internal/templates/<int:template_id>/full-print` - `GET`
* `/internal/templates/length` - `GET`
* `/internal/templates/<int:template_id>` - `DELETE`

**Credentials API**
* `/internal/credentials` - `GET`
* `/internal/credentials/renew` - `PUT`

**Jobtree API**
* `/internal/jobtree` - `GET`

**Plugins API**
* `/internal/plugins` - `GET`
* `/internal/plugins/<plugin_name>` - `GET`

**Other API**
* `/internal/jobs/actions` - `GET`
* `/internal/load` - `GET`
* `/internal/runfile` - `GET`
* `/ping` - `GET`

---

### About `gui` Web Server

The `gui` server is what the user interacts with, it handles the rendering and serving of HTML, CSS, JS files. It also manages authentication states and other security features. The `gui` server has the following view routes implemented as well as the API routes, these APIs are accessible by the user with a valid token. Refer to the documentation for more information.

**View Routes**
* `/login`
* `/logout`
* `/`
* `/config`
* `/create`
* `/create/runfile`
* `/templates`
* `/jobs`
* `/jobs/<int:job_id>`
* `/jobs/<int:job_id>/export`
* `/jobs/<int:job_id>/edit`
* `/job/<int:job_id>/browse`
* `/job/<int:job_id>/browse/<path:path>`
* `/jobs/<int:job_id>/subjobs`
* `/jobs/<int:job_id>/subjobs/<int:subjob_id>`
* `/credentials`
* `/plugins`
* `/plugin/<plugin_name>`
* `/logs`
* `/storage`
* `/storage/<path:path>`
* `/cli`
* `connect` - `Websocket`
* `pty-input` - `Websocket`
* `resize` - `Websocket`
* `/shutdown` 

**Web APIs**

* Token
    * `/token` - `POST`
* Job Web API
    * `/api/jobs/<int:job_id>` - `GET`
    * `/api/jobs/<int:job_id>/<attribute>` - `GET`
    * `/api/jobs/<int:job_id>/full-print` - `GET`
    * `/api/jobs/create` - `POST`
    * `/api/jobs/<int:job_id>/copy` - `PUT`
    * `/api/jobs/<int:job_id>/<action>` - `PUT`
    * `/api/jobs/<int:job_id>` - `DELETE`
    * `/api/jobs/<int:job_id>/pin` - `GET`
    * `/api/jobs/<int:job_id>/unpin` - `GET`
* Subjobs Web API
    * `/api/jobs/<int:job_id>/subjobs` - `GET`
    * `/api/jobs/<int:job_id>/subjobs/<int:subjob_id>` - `GET`
    * `/api/jobs/<int:job_id>/subjobs/<int:subjob_id>/<attribute>` - `GET`
    * `/api/jobs/<int:job_id>/subjobs/<int:subjob_id>/full-print` - `GET`
    * `/api/jobs/<int:job_id>/subjobs/<int:subjob_id>/copy` - `PUT`
* Jobs Web API
    * `/api/jobs` - `GET`
    * `/api/jobs/statistics` - `GET`
    * `/api/jobs/incomplete_ids` - `GET`
* Config Web API
    * `/api/config` - `GET`
    * `/api/config/<section>` - `GET`
* Tempaltes Web API
    * `/api/templates` - `GET`
    * `/api/templates/<int:template_id>/full-print` - `GET`
    * `/api/templates/<int:template_id>` - `DELETE`
* Credentails Web API
    * `/api/credentials` - `GET`
    * `/api/credentials/renew` - `PUT`
* Jobtree Web API
    * `/api/jobtree` - `GET`
* Plugins Web API
    * `/api/plugins` - `GET`
    * `/api/plugins/<plugin_name>` - `GET`

All the communication with `gui` Web API will happen in JSON format with a valid token in request headers as `X-Access-Token`. 

---

### About Web CLI

The usual pattern of starting the Ganga GUI is by starting it using `--webgui` flag while starting Ganga or `start_gui` function in the Ganga CLI. This method starts a Ganga session first and then the `internal` server and then the `gui` server. 

In the case when the user wants to access the Ganga CLI on their browser over the network. They will need to start the Ganga using a special binary called `ganga-gui`. It takes `host`, `port`, `internal-port`, `--quite` and `ganga-args` as the command line arguments. What this script does is, it starts the `gui` server first and then when the first HTTP request is made to the server, it start a Ganga session in a pseudo-terminal with the `--webgui` flag (in this case the webgui flag just starts the internal server). With this approach, the `gui` server has access to the pseudo terminal's file descriptor in which the Ganga session is running which is then used to interact with Ganga.

---

### TODO

- [ ] Add ability to auto search free port for starting the servers.
- [ ]  Write internal and user documentation
- [ ] Minor ease of use changes like sorting of tables, etc.

---

__The start page for this project can be accessed [here](https://github.com/ganga-devs/ganga/issues/1663).__

### CODE REVIEW

**Comment by egede** [comment]
It would be good to get this PR to a state where it can be merged (so no new features, but fixing the remaining test failures).

**Comment by varunbankar** [comment]
> It would be good to get this PR to a state where it can be merged (so no new features, but fixing the remaining test failures).

Yes, I'll get onto that. Next 2 days I'll plan and split this PR into multiple smaller ones and fix the test cases.

File: ganga/GangaGUI/api/__init__.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+cli = sys.modules['flask.cli']
+cli.show_server_banner = lambda *x: None
+
+# Only error cases are logged
+log = logging.getLogger('werkzeug')
+log.setLevel(logging.ERROR)
+
+# Flask app for APIs
+internal = Flask(__name__)
+internal.config['SECRET_KEY'] = "f3aa26t8b537abf6ee6305eefe2wr10a" # TODO Generate secret key
```
Comment (egede) [inline ganga/GangaGUI/api/__init__.py:15]: What is this SECRET_KEY used for? As it sits in the publicly viewable code on GitHub, it is not very secret!

File: ganga/GangaGUI/ganga-gui [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+# Parse command line arguments
+web_cli_host = args.host
+web_cli_port = int(args.port)
+internal_server_port = int(args.internal_port)
+ganga_args = args.ganga_args
+log_output = not args.quite
+
+# Start the web cli - this will also start a Ganga Session
+start_web_cli(host=web_cli_host, port=web_cli_port, internal_port=internal_server_port, ganga_args=ganga_args, log_output=log_output)
```
Comment (egede) [inline ganga/GangaGUI/ganga-gui:31]: This script should be moved to the `bin` directory where the main binary for starting Ganga is also located.

File: ganga/GangaGUI/gui/config.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# Make GUI essential folders
+for dir_name in ["upload", "logs", "storage"]:
+    dir_path = os.path.join(gui_dir, dir_name)
+    if not os.path.exists(dir_path):
+        os.makedirs(dir_path)
+
+
+# TODO Look into ways to generate secret_key without much input from user side - .gangarc?
+class Config(object):
+    SECRET_KEY = os.environ.get("SECRET_KEY") or "f3aa26t8b537abf6ee6305eefea0a10a"
```
Comment (egede) [inline ganga/GangaGUI/gui/config.py:25]: How is this related to the SECRET_KEY above?

File: ganga/GangaGUI/gui/routes.py [hunk 4]
```diff
@@ -1,71 +1,1018 @@
...
+
+# Colors showed for different job statuses in the GUI based on Bootstrap CSS
+status_color = {
+    "new": "info",
+    "completed": "success",
+    "failed": "danger",
+    "running": "primary",
+    "submitted": "secondary",
+    "killed": "warning"
+}
```
Comment (egede) [inline ganga/GangaGUI/gui/routes.py:89]: What happens if the job has a status that is not in this table? We have two other statuses that a job can be in for a short time (`submitting` and `completing`).

File: ganga/GangaGUI/gui/static/js/main.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+// Global variables
+let notificationCount = 0;
+const statusColor = {
+    'new': 'info',
+    'completed': 'success',
+    'failed': 'danger',
+    'running': 'primary',
+    'submitted': 'secondary',
+    'killed': 'warning'
+};
```
Comment (egede) [inline ganga/GangaGUI/gui/static/js/main.js:26]: Why is this duplicated in different places?

**Review by egede — APPROVED** [review]
This all looks good. There are only a few questions, but no show stoppers.

File: ganga/GangaGUI/api/__init__.py (hunk 1 above)
Comment (varunbankar) [reply ganga/GangaGUI/api/__init__.py:15]: This SECRET_KEY is used by the `internal` RESTful server to which the `gui` server communicates in order to get data from Ganga, The SECRET_KEY is used by the server for validating the integrity of the requests if it had sent some encoded data in it (eg. the token-based authentication uses this SECREY_KEY to encrypt and decrypt the tokens). I will merge this PR and create a new one where I will fix this secret key by randomly creating it for the user.

File: ganga/GangaGUI/ganga-gui (hunk 2 above)
Comment (varunbankar) [reply ganga/GangaGUI/ganga-gui:31]: Okay, I will shift this.

File: ganga/GangaGUI/gui/config.py (hunk 3 above)
Comment (varunbankar) [reply ganga/GangaGUI/gui/config.py:25]: This is the SECRET_KEY used by the `gui` server which is what the user interacts with.

File: ganga/GangaGUI/gui/routes.py (hunk 4 above)
Comment (varunbankar) [reply ganga/GangaGUI/gui/routes.py:89]: The status will be shown, it just won't have a color. It will be shown a plain text instead of green background (which you see in completed jobs) or such.

File: ganga/GangaGUI/gui/static/js/main.js (hunk 5 above)
Comment (varunbankar) [reply ganga/GangaGUI/gui/static/js/main.js:26]: The above one is used by the server when rendering HTML pages whereas this one is used on the client-side (the browser). It is the JavaScript version.
